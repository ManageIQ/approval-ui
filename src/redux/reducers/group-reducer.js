import {
  FETCH_RBAC_GROUPS,
  FETCH_GROUP_NAMES
} from '../action-types';

// Initial State
export const groupsInitialState = {
  groups: [],
  isLoading: false
};

const setLoadingState = state => ({ ...state, isLoading: true });
const setGroups = (state, { payload }) => ({ ...state, groups: payload, isLoading: false });
const setGroupNames = (state, { payload }) => ({ ...state, groupNames: payload, isLoading: false });

export default {
  [`${FETCH_RBAC_GROUPS}_PENDING`]: setLoadingState,
  [`${FETCH_RBAC_GROUPS}_FULFILLED`]: setGroups,
  [`${FETCH_GROUP_NAMES}_PENDING`]: setLoadingState,
  [`${FETCH_GROUP_NAMES}_FULFILLED`]: setGroupNames
};
