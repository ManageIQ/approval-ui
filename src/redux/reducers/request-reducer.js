import {
  FETCH_REQUEST,
  FETCH_REQUEST_CONTENT,
  FETCH_REQUESTS,
  EXPAND_REQUEST
} from '../../redux/action-types';

// Initial State
export const requestsInitialState = {
  requests: {
    data: [],
    meta: {
      count: 0,
      limit: 10,
      offset: 0
    }
  },
  filterValue: '',
  isRequestDataLoading: false,
  expandedRequests: [],
  selectedRequest: {
    metadata: {
      user_capabilities: {}
    },
    requests: []
  }
};

const setLoadingState = state => ({ ...state, isRequestDataLoading: true, expandedRequests: []});
const setRequests = (state, { payload }) => ({ ...state, requests: payload, isRequestDataLoading: false });
const selectRequest = (state, { payload }) => ({ ...state, selectedRequest: payload, isRequestDataLoading: false });
const setRequestContent = (state, { payload }) => ({ ...state, requestContent: payload, isRequestDataLoading: false });
const setexpandRequest = (state, { payload }) => ({ ...state, expandedRequests: [ ...state.expandedRequests, payload ]});

export default {
  [`${FETCH_REQUESTS}_PENDING`]: setLoadingState,
  [`${FETCH_REQUESTS}_FULFILLED`]: setRequests,
  [`${FETCH_REQUEST}_PENDING`]: setLoadingState,
  [`${FETCH_REQUEST}_FULFILLED`]: selectRequest,
  [`${FETCH_REQUEST_CONTENT}_PENDING`]: setLoadingState,
  [`${FETCH_REQUEST_CONTENT}_FULFILLED`]: setRequestContent,
  [EXPAND_REQUEST]: setexpandRequest
};
