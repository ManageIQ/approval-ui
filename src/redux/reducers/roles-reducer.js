import { SET_USER_ROLES } from '../action-types';
import { isApprovalAdmin } from '../../helpers/shared/helpers';

export const rolesInitialState = {
  userRoles: [],
  approvalAdmin: false
};
const setUserRoles = (state, { payload }) => ({ ...state, userRoles: payload, approvalAdmin: isApprovalAdmin(payload) });

export default {
  [SET_USER_ROLES]: setUserRoles
};
