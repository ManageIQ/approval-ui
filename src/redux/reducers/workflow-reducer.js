import {
  FETCH_WORKFLOW,
  FETCH_WORKFLOWS,
  SORT_WORKFLOWS,
  SET_FILTER_WORKFLOWS,
  MOVE_SEQUENCE,
  UPDATE_WORKFLOW
} from '../../redux/action-types';

// Initial State
export const workflowsInitialState = {
  workflows: {
    data: [],
    meta: {
      count: 0,
      limit: 50,
      offset: 0
    }
  },
  workflow: {},
  filterValue: '',
  isLoading: false,
  isRecordLoading: false,
  isUpdating: 0,
  sortBy: {
    index: 2,
    property: 'sequence',
    direction: 'asc'
  }
};

const setLoadingState = state => ({ ...state, isLoading: true });
const setRecordLoadingState = state => ({ ...state, isRecordLoading: true });
const setWorkflows = (state, { payload }) => ({ ...state, workflows: payload, isLoading: false });
const selectWorkflow = (state, { payload }) => ({ ...state, workflow: payload, isRecordLoading: false });
const setSortWorkflows = (state, { payload }) => ({
  ...state,
  sortBy: payload,
  workflows: {
    ...state.workflows,
    meta: {
      ...state.workflows.meta,
      offset: 0
    }
  }
});
const setFilterValue = (state, { payload }) => ({
  ...state,
  filterValue: payload,
  workflows: {
    ...state.workflows,
    meta: {
      ...state.workflows.meta,
      offset: 0
    }
  }
});

const tranformWf = (wf, id, sequence, isMovingUp) => {
  if (wf.id === id) {
    return { ...wf, sequence };
  }

  if (wf.sequence === sequence && isMovingUp) {
    return { ...wf, sequence: wf.sequence - 1 };
  }

  if (wf.sequence === sequence && !isMovingUp) {
    return { ...wf, sequence: wf.sequence + 1 };
  }

  return wf;
};

const moveSequence = (state, { payload }) => {
  const originalProcess = state.workflows.data.find(({ id }) => id === payload.id);

  const isMovingUp = payload?.sequence > originalProcess?.sequence;

  return ({
    ...state,
    workflows: {
      ...state.workflows,
      data: state.workflows.data
      .map(wf => tranformWf(wf, payload.id, payload.sequence, isMovingUp))
      .sort((a, b) => state.sortBy.direction === 'asc' ? a.sequence - b.sequence : b.sequence - a.sequence)
    }
  });
};

const setUpdatingWorkflow = (state) => ({
  ...state,
  isUpdating: state.isUpdating + 1
});
const finishUpdatingWorkflow = (state) => ({
  ...state,
  isUpdating: state.isUpdating - 1
});

export default {
  [`${FETCH_WORKFLOWS}_PENDING`]: setLoadingState,
  [`${FETCH_WORKFLOWS}_FULFILLED`]: setWorkflows,
  [`${FETCH_WORKFLOW}_PENDING`]: setRecordLoadingState,
  [`${FETCH_WORKFLOW}_FULFILLED`]: selectWorkflow,
  [`${UPDATE_WORKFLOW}_PENDING`]: setUpdatingWorkflow,
  [`${UPDATE_WORKFLOW}_FULFILLED`]: finishUpdatingWorkflow,
  [`${UPDATE_WORKFLOW}_REJECTED`]: finishUpdatingWorkflow,
  [SORT_WORKFLOWS]: setSortWorkflows,
  [SET_FILTER_WORKFLOWS]: setFilterValue,
  [MOVE_SEQUENCE]: moveSequence
};
