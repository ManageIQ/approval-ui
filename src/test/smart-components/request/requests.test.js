import React from 'react';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store' ;
import { shallowToJson } from 'enzyme-to-json';
import promiseMiddleware from 'redux-promise-middleware';
import { IntlProvider } from 'react-intl';
import Requests from '../../../smart-components/request/requests';
import { requestsInitialState } from '../../../redux/reducers/request-reducer';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications';

const ComponentWrapper = ({ store, initialEntries = [ '/requests' ], children }) => (
  <Provider store={ store }>
    <MemoryRouter initialEntries={ initialEntries }>
      <IntlProvider locale="en">
        { children }
      </IntlProvider>
    </MemoryRouter>
  </Provider>
);

describe('<Requests />', () => {

  let initialProps;
  const middlewares = [ thunk, promiseMiddleware(), notificationsMiddleware() ];
  let mockStore;
  let initialState;

  beforeEach(() => {
    initialProps = {};
    mockStore = configureStore(middlewares);
    initialState = { requestReducer: { ...requestsInitialState, isLoading: false }};
  });

  it('should render correctly', () => {
    const store = mockStore(initialState);
    const wrapper = shallow(<ComponentWrapper store={ store }><Requests { ...initialProps } /></ComponentWrapper>);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should render correctly in loading state', () => {
    const store = mockStore(initialState);
    const wrapper = shallow(<ComponentWrapper store={ store }><Requests { ...initialProps } isLoading /></ComponentWrapper>);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
