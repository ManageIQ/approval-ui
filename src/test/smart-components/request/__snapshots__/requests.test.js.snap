// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Requests /> should render correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MemoryRouter
    initialEntries={
      Array [
        "/requests",
      ]
    }
  >
    <IntlProvider
      locale="en"
    >
      <Connect(Requests) />
    </IntlProvider>
  </MemoryRouter>
</Provider>
`;

exports[`<Requests /> should render correctly in loading state 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MemoryRouter
    initialEntries={
      Array [
        "/requests",
      ]
    }
  >
    <IntlProvider
      locale="en"
    >
      <Connect(Requests)
        isLoading={true}
      />
    </IntlProvider>
  </MemoryRouter>
</Provider>
`;
